src/Microsoft.Azure.WebJobs.Extensions.Kafka/Listeners/KafkaTargetScaler.cs

```diff
@@ -0,0 +1,158 @@
+            if (metrics == null || DateTime.UtcNow - metrics.Timestamp > metricsTimeOut)
+            {
+                metrics = await this.metricsProvider.GetMetricsAsync();
+                this.logger.LogInformation($"Calculating metrics as last calculated don't exist or were stored 2 minutes ago.");
+            }
+            return metrics;
+        }
+
+        internal TargetScalerResult GetScaleResultInternal(TargetScalerContext context, KafkaTriggerMetrics metrics)
+        {
+            var totalLag = metrics.TotalLag;
+            var partitionCount = metrics.PartitionCount;
+
+            // Since Kafka Extension supports only Premium plan for run time based scaling,
+            // the targetWorkerCount is set to 1 even when the totalLag is 0.
+            // This can be changed to 0 when the extension supports all plans.
+            if (totalLag == 0)
+            {
+                return new TargetScalerResult
+                {
+                    TargetWorkerCount = 1
+                };
+            }
+
+            var targetConcurrency = GetConcurrency(context, this.lagThreshold);
+            
+            int targetWorkerCount = (int) Math.Ceiling(totalLag / (decimal) targetConcurrency);
+
+            targetWorkerCount = ValidateWithPartitionCount(targetWorkerCount, partitionCount);
+            targetWorkerCount = ThrottleResultIfNecessary(targetWorkerCount);
+            if (GetChangeInWorkerCount(targetWorkerCount) > 0)
+            {
+                this.lastScaleUpTime = DateTime.UtcNow;
+            }
+
+            this.logger.LogInformation($"Total Lag: {totalLag}, concurrency: {targetConcurrency} TargetWorkerCount: {targetWorkerCount}. For the topic {this.topicName}, consumer group {consumerGroup}.");   
+
+            return new TargetScalerResult
+            {
+                TargetWorkerCount = targetWorkerCount
+            };
+        }
+
+        internal int GetConcurrency(TargetScalerContext context, long lagThreshold)
+        {
+            // If dynamicConcurrencyEnabled is set to true, target concurrency is
+            // set to instanceConcurrency value, else it is set to lagThreshold
+            // (default value = 1000).
+            int targetConcurrency = context.InstanceConcurrency ?? (int) lagThreshold;
+            if (targetConcurrency < 1)
+            {
+                throw new ArgumentException("Target concurrency must be larger than 0.");
+            }
+            return targetConcurrency;
+        }
+
+        internal int ValidateWithPartitionCount(int targetWorkerCount, long partitionCount)
+        {
+            // Limit targetWorkerCount to number of partitions.
+            if (targetWorkerCount > partitionCount)
+            {
+                targetWorkerCount = (int) partitionCount;
+            }
+
+            return targetWorkerCount;
+        }
+
+        internal int ThrottleResultIfNecessary(int targetWorkerCount)
```

Comment:
Is this needed? I'm thinking if this method could cause less than ideal throttle down for the function app

Reply:
Scaling up is an expensive process while compared to scaling down. Idea here being - any method, ideal or not, that prevents immediate scaling down right after a scale up would be helpful. The most basic thing to do is restricting scale down right after scaling up for a certain amount of time.